compiler/rustc_middle/src/mir/interpret/allocation.rs

```diff
@@ -21,6 +22,98 @@ use super::{
 };
 use crate::ty;
 
+/// Functionality required for the bytes of an `Allocation`.
+pub trait AllocBytes:
+    Clone + core::fmt::Debug + Eq + PartialEq + PartialOrd + Ord + core::hash::Hash
```

Comment:
If you make this also have `Deref<Target = [u8]>`, you can remove everything but `adjust_to_align`, `from_bytes` and `uninit` and should have to adjust much less code as it would pick up the slice methods that `Box<[u8]>` also exposes via `Deref`.

Reply:
Ah good point! I was hoping to keep `AllocBytes` more general though, so I can use it to represent foreign memory in Miri, that isn't represented by a `[u8]`. Given that, do you think it's ok to keep these extra functions in `AllocBytes`?

Reply:
What kind of memory are you thinking of? You already support `get_slice_from_range`, which returns a `&[u8]`. Most of what I'm proposing is to get rid of that method and require the implementor of `AllocBytes` to put the logic into the `Deref::deref` method instead. Or are you planning to change the return type of the `get_slice` methods?

Reply:
Ah, if we can make a custom `get_slice` method with the same return type then that should work. In my implementation these methods still return `&[u8]` -- my current representation of foreign memory is represented as a pair of machine address and length, and the `get_slice` methods build the slice from `std::slice::from_raw_parts`.

Reply:
yea, so you could implement `Deref` and `DerefMut` for your foreign memory representation and have the `deref` and `deref_mut` methods use `from_raw_parts`. Then you can remove the `get_slice*` methods and others from the trait, as you get them for free once `Deref` is a super trait

Reply:
Oh I see -- that's smart, thanks for the suggestion! (Will push these changes soon)

Reply:
I agree the trait has a lot more methods than I was expecting so this sounds good.